dali/pipeline/operators/operator.h

```diff
@@ -176,6 +176,16 @@ class Operator : public OperatorBase {
     SetupSharedSampleParams(ws);
 
     for (int i = 0; i < input_sets_; ++i) {
+      if (std::is_same<Backend, GPUBackend>::value) {
+        // Before we start working on the next input set, we need
+        // to wait until the last one is finished. Otherwise for some ops
+        // we risk overwriting data used by the kernel called for previous
+        // image. Doing it for all ops is a compromise between performance
+        // (which should not be greatly affected) and robustness (guarding
```

Comment:
Do we have any measurement showing how much the performance is affected?

Reply:
In the current benchmark cases it is exactly 0 because we don't use multiple input sets there. For a general case I would say it should be pretty minimal - waiting would happen in a side thread anyway, the only performance related problem is that with that wait you would not launch the kernel for the next set right away, but instead you would need to perform the data dependent setup in the critical path. But that setup is really fast for all operators anyway, so it should not matter.

Reply:
And in most cases where such setup exists, you actually need to do the synchronization already, so the impact should be even lower.

Reply:
Sounds reasonable